src/python/pants/backend/python/interpreter_cache.py

```diff
@@ -134,22 +139,20 @@ def _setup_interpreter(self, interpreter, cache_target_path):
   def _setup_cached(self, filters=()):
     """Find all currently-cached interpreters."""
     for interpreter_dir in os.listdir(self._cache_dir):
-      path = os.path.join(self._cache_dir, interpreter_dir)
-      if os.path.isdir(path):
-        pi = self._interpreter_from_path(path, filters=filters)
-        if pi:
-          logger.debug('Detected interpreter {}: {}'.format(pi.binary, str(pi.identity)))
-          yield pi
+      pi = self._interpreter_from_relpath(interpreter_dir, filters=filters)
+      if pi:
+        logger.debug('Detected interpreter {}: {}'.format(pi.binary, str(pi.identity)))
+        yield pi
 
   def _setup_paths(self, paths, filters=()):
     """Find interpreters under paths, and cache them."""
     for interpreter in self._matching(PythonInterpreter.all(paths), filters=filters):
       identity_str = str(interpreter.identity)
-      cache_path = os.path.join(self._cache_dir, identity_str)
-      pi = self._interpreter_from_path(cache_path, filters=filters)
+      pi = self._interpreter_from_relpath(identity_str, filters=filters)
       if pi is None:
+        cache_path = os.path.join(self._cache_dir, identity_str)
         self._setup_interpreter(interpreter, cache_path)
```

Comment:
It would be a bit more symmetric on the whole and less error-prone here to make `_setup_interpreter` take a relpath (`identity_str`) and have it do the final cache path construction.